did_core/did_parser_nom/src/did_url/parsing.rs

```diff
@@ -0,0 +1,167 @@
+        char('='),
+        alt((take_while1(|c| !"&#?".contains(c)), success(""))),
+    ))(input)?;
+
+    cut(all_consuming(take_while1(is_query_char)))(key)?;
+    if !value.is_empty() {
+        cut(all_consuming(take_while1(is_query_char)))(value)?;
+    }
+
+    Ok((remaining, (key, value)))
+}
+
+fn query_parser(input: &str) -> IResult<&str, Vec<(&str, &str)>> {
+    separated_list0(one_of("&?"), query_key_value_pair)(input)
+}
+
+fn parse_did_ranges_with_empty_allowed(input: &str) -> IResult<&str, DidRanges> {
+    alt((
+        parse_did_ranges,
+        success(Default::default()), // Relative DID URL
+    ))(input)
+}
+
+// did-url-remaining = path-abempty [ "?" query ] [ "#" fragment ]
+fn parse_url_part(input: &str) -> IResult<&str, UrlPart> {
+    let (remaining, path) = path_abempty(input)?;
+    let (remaining, queries) = opt(preceded(tag("?"), cut(query_parser)))(remaining)?;
+    let (remaining, fragment) =
+        opt(preceded(tag("#"), cut(all_consuming(fragment_parser))))(remaining)?;
+    Ok((remaining, (path, queries, fragment)))
+}
+
+fn to_did_url_ranges(
+    id_range: Option<DidRange>,
+    (path, queries, fragment): UrlPart,
+) -> (
+    Option<DidRange>,
+    HashMap<DidRange, DidRange>,
+    Option<DidRange>,
+) {
+    let id_end = id_range.unwrap_or_default().end;
+    let path_range = if path.is_empty() {
+        None
+    } else {
+        let path_start = id_end;
+        let path_end = path_start + path.len();
+        Some(path_start..path_end)
+    };
+
+    let mut current_last_position = path_range
+        .clone()
+        .map(|range| range.end + 1)
+        .unwrap_or(id_end + 1);
+
+    let mut query_map = HashMap::<DidRange, DidRange>::new();
+    for (key, value) in queries.unwrap_or_default() {
+        let key_start = current_last_position;
+        let key_end = key_start + key.len();
+        let value_start = key_end + 1;
+        let value_end = value_start + value.len();
+        current_last_position = value_end + 1;
+        query_map.insert(key_start..key_end, value_start..value_end);
+    }
+
+    let fragment_range = fragment.map(|fragment| {
+        let fragment_end = fragment.len() + current_last_position;
+        current_last_position..fragment_end
+    });
+
+    (path_range, query_map, fragment_range)
+}
+
+fn validate_result_not_empty(url_part: &UrlPart, did_ranges: &DidRanges) -> Result<(), ParseError> {
+    if (url_part, did_ranges) == (&Default::default(), &Default::default()) {
+        Err(ParseError::InvalidInput("Invalid input"))
+    } else {
+        Ok(())
+    }
+}
+
+// did-url = did path-abempty [ "?" query ] [ "#" fragment ]
+pub fn parse_did_url(did_url: String) -> Result<DidUrl, ParseError> {
+    if did_url.is_empty() {
+        return Err(ParseError::InvalidInput("Empty input"));
+    }
+
+    let (remaining, did_ranges) = parse_did_ranges_with_empty_allowed(&did_url)?;
+
+    let (_, url_part) = all_consuming(parse_url_part)(remaining)?;
```

Comment:
I have a hunch this could be prettier if we don't deal with `DidRanges`, just in terms of raw parsers... but I am speaking bit out of my ballpark, not sure what challenges I am missing.

Just a hunch, if `parse_did_ranges` would return `IResult<&str, DidRanges>`, could't we instead have here something like 
```
let (did_parts_tuple, did_url_parts_tuple) = tuple((
  parse_did_ranges,
  parse_url_part
))?;
```
?

Reply:
I am not sure I understand what you mean or are going for, but I strongly disagree with a parser returning ranges.

Reply:
I had feeling like this could be refactored, but wasn't' quite sure about details. If this didn't ring any bells, chances are this has no merit. Not a big deal anyway, the code is well readable as it is.